src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs

```diff
@@ -662,6 +667,112 @@ private void OnHostLaunched(HostProviderEventArgs e)
         }
 
+        /// <summary>
+        /// Check if the testhost.dll is compatible with the chosen muxer, if not update with the one close to it.
+        /// This cover the scenario where we're switching to a different architecture than the sdk self(current process) and
+        /// we end to pick up testhost.dll from the sdk folder.
+        /// testhost.dll is compiled with the architecture metadata and will fail if used from a muxer with different architecture.
+        /// </summary>
+        /// <param name="testHostPath">Path of the testhost.dll</param>
+        /// <param name="muxerPath">Path of the muxer</param>
+        /// <param name="targetArchitecture">target architecture</param>
+        /// <returns>return the path of testhost.dll to use</returns>
+        private string CheckTestHostDllArchitecture(string testHostPath, string muxerPath, PlatformArchitecture targetArchitecture)
+        {
+            // By default compatibility is true, in case of unknown architecture we'll fail.
+            bool isCompatibleTestHostDll = true;
+            using (Stream stream = this.fileHelper.GetStream(testHostPath, FileMode.Open, FileAccess.Read))
+            using (PEReader peReader = new PEReader(stream))
+            {
+                switch (peReader.PEHeaders.CoffHeader.Machine)
+                {
+                    case Machine.IA64:
+                    case Machine.Amd64:
+                        isCompatibleTestHostDll = targetArchitecture == PlatformArchitecture.X64;
+                        break;
+                    case Machine.Arm64:
+                        isCompatibleTestHostDll = targetArchitecture == PlatformArchitecture.ARM64;
+                        break;
+                    case Machine.Arm:
+                        isCompatibleTestHostDll = targetArchitecture == PlatformArchitecture.ARM;
+                        break;
+                    case Machine.I386:
+                        var headers = peReader.PEHeaders;
+                        var characteristics = headers.CoffHeader.Characteristics;
+                        var corflags = headers.CorHeader.Flags;
+
+                        // If requires 32bit or if R2R
+                        if ((corflags & CorFlags.Requires32Bit) != 0 || ((corflags & CorFlags.ILOnly) == 0))
+                        {
+                            isCompatibleTestHostDll = targetArchitecture == PlatformArchitecture.X86;
+                        }
+                        else
+                        {
+                            EqtTrace.Error("DotnetTestHostmanager: AnyCPU testhost.dll provided {0}", testHostPath);
+                            return testHostPath;
+                        }
+
+                        break;
+                    default:
+                        EqtTrace.Error("DotnetTestHostmanager: Unknown architecture {0}", peReader.PEHeaders.CoffHeader.Machine);
```

Comment:
Nit: Adding the end of sentence period.

```suggestion
                        EqtTrace.Error("DotnetTestHostmanager: Unknown architecture {0}.", peReader.PEHeaders.CoffHeader.Machine);
```